src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -87,13 +89,16 @@ public override BoundNode VisitTryStatement(BoundTryStatement node)
             return null;
         }
 
-        protected virtual void VisitTryBlock(BoundStatement tryBlock, BoundTryStatement node, ref TLocalState tryState)
+        protected virtual void VisitTryBlock(BoundBlock tryBlock, BoundTryStatement node, ref TLocalState tryState)
         {
+            var oldPending = SavePending();
             VisitStatement(tryBlock);
+            RestorePending(oldPending);
```

Comment:
>RestorePending(oldPending); [](start = 12, length = 27)

It feels like instead of using this helper, we need to have a helper that only resolves branches to  labels in _labelsSeen. Same for catch blocks. And we probably need to make similar changes in the base class too.

Reply:
I mean we want to resolve pending branches, but we don't want to restore old pending branches just yet. We want to stick this pending branches and only with pending branches originated from within the ```try``` and jumping out of it.

---
In reply to: [151229092](https://github.com/dotnet/roslyn/pull/23172#discussion_r151229092) [](ancestors = 151229092)

Reply:
It could be useful here to separate `RestorePending` into resolving of branches and merging with older state steps. It would be very uncommon for these steps to be used separately and could be source of bugs.
Also, something with the diff as Codeflow shows it. Perhaps caught one of earlier commits, but this was later fixed. We restore `oldPending` later in the method.

So, the restoring too early is fixed. Splitting the `RestorePending`  into two stages - it think it is not worth it.

---
In reply to: [151236006](https://github.com/dotnet/roslyn/pull/23172#discussion_r151236006) [](ancestors = 151236006,151229092)

Reply:
I will add more comments in this method to make it easier to understand what is happening.

---
In reply to: [151302455](https://github.com/dotnet/roslyn/pull/23172#discussion_r151302455) [](ancestors = 151302455,151236006,151229092)